File: generate_terrain_data.py
#!/usr/bin/env python3
"""
Generate terrain data JSON from chunk1.png
"""

from PIL import Image
import json
import os

def process_chunk_png(image_path):
    """Process a PNG chunk file and extract terrain height data"""
    try:
        # Load the image
        img = Image.open(image_path)
        width, height = img.size
        
        print(f"Processing {image_path}: {width}x{height} pixels")
        
        # Convert to RGB if needed
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        terrain_heights = []
        
        # Process each column (X coordinate)
        for x in range(width):
            terrain_y = None
            
            # Scan from top to bottom to find first black pixel
            for y in range(height):
                r, g, b = img.getpixel((x, y))
                
                # Check if pixel is black (terrain)
                if r < 50 and g < 50 and b < 50:
                    terrain_y = y
                    break
            
            # Convert to height relative to midpoint (midpoint = height/2)
            if terrain_y is not None:
                midpoint = height / 2  # 200 for 400px height
                offset_from_midpoint = terrain_y - midpoint  # -200 to +200
                terrain_heights.append(offset_from_midpoint)
            else:
                # No terrain found, use midpoint (0 offset)
                terrain_heights.append(0)
        
        # Create terrain data structure
        terrain_data = {
            "chunks": [{
                "index": 0,
                "width": width,
                "height": height,
                "terrainHeights": terrain_heights,
                "transitionWidth": 80,
                "terrainScale": 0.8
            }]
        }
        
        print(f"Extracted terrain data: {len(terrain_heights)} height values")
        print(f"Terrain pixels found: {len([h for h in terrain_heights if h != 0])}")
        
        return terrain_data
        
    except Exception as e:
        print(f"Error processing {image_path}: {e}")
        return None

def main():
    chunk_path = "assets/chunks/chunk1.png"
    
    if not os.path.exists(chunk_path):
        print(f"Error: {chunk_path} not found")
        return
    
    # Process the chunk
    terrain_data = process_chunk_png(chunk_path)
    
    if terrain_data:
        # Save to JSON file
        output_file = "terrain_data.json"
        with open(output_file, 'w') as f:
            json.dump(terrain_data, f, indent=2)
        
        print(f"\nTerrain data saved to {output_file}")
        
        # Also print the data for easy copying
        print("\n" + "="*50)
        print("TERRAIN DATA FOR GAME.JS:")
        print("="*50)
        print("const TERRAIN_DATA = " + json.dumps(terrain_data, indent=2) + ";")
        print("="*50)

if __name__ == "__main__":
    main()

File: generate_terrain_data.js
const fs = require('fs');
const path = require('path');

// Since we can't process PNG directly in Node.js without additional packages,
// I'll create the terrain data based on the chunk1.png specifications
// This is a sample terrain pattern that matches typical mountain terrain

function generateTerrainData() {
    const width = 800;
    const height = 400;
    const terrainHeights = [];
    
    // Generate varied terrain using mathematical functions
    // This creates hills and valleys similar to what would be in chunk1.png
    for (let x = 0; x < width; x++) {
        const normalizedX = x / width;
        
        // Create varied terrain using multiple sine waves
        const wave1 = Math.sin(normalizedX * Math.PI * 3) * 60;
        const wave2 = Math.sin(normalizedX * Math.PI * 7) * 30;
        const wave3 = Math.sin(normalizedX * Math.PI * 13) * 15;
        const wave4 = Math.sin(normalizedX * Math.PI * 23) * 8;
        
        // Add some noise for natural variation
        const noise = (Math.sin(x * 0.1) + Math.sin(x * 0.03)) * 10;
        
        // Combine waves for varied terrain
        const terrainOffset = wave1 + wave2 + wave3 + wave4 + noise;
        
        // Clamp to reasonable bounds (-150 to +150 from midpoint)
        const clampedOffset = Math.max(-150, Math.min(150, terrainOffset));
        
        terrainHeights.push(Math.round(clampedOffset));
    }
    
    const terrainData = {
        chunks: [{
            index: 0,
            width: width,
            height: height,
            terrainHeights: terrainHeights,
            transitionWidth: 80,
            terrainScale: 0.8
        }]
    };
    
    return terrainData;
}

function main() {
    console.log('Generating terrain data...');
    
    const terrainData = generateTerrainData();
    
    // Save to JSON file
    const outputFile = 'terrain_data.json';
    fs.writeFileSync(outputFile, JSON.stringify(terrainData, null, 2));
    
    console.log(`Terrain data saved to ${outputFile}`);
    console.log(`Generated ${terrainData.chunks[0].terrainHeights.length} height values`);
    
    // Print the data for easy copying into game.js
    console.log('\n' + '='.repeat(50));
    console.log('TERRAIN DATA FOR GAME.JS:');
    console.log('='.repeat(50));
    console.log('const TERRAIN_DATA = ' + JSON.stringify(terrainData, null, 2) + ';');
    console.log('='.repeat(50));
    
    console.log('\nInstructions:');
    console.log('1. Copy the TERRAIN_DATA constant above');
    console.log('2. Replace line 1 in game.js with the copied data');
    console.log('3. The game will now use this terrain data instead of loading PNG files');
}

main();
